File: linear_regression/LinearRegression.cpp
#include "LinearRegression.hpp"

/**
 * Implementation of a simple linear regression model in c++.
 * formula: y = Wx + b
 */

LinearRegression::LinearRegression() : a(0.01) {}

LinearRegression::~LinearRegression() {}

/**
 *  Calculate mean squared error 
 *  @param yTrue: true values
 *  @param yPred: predicted values
 *  @return: mean squared error
 */
double LinearRegression::meanSquaredError(const std::vector<double>& yTrue, const std::vector<double>& yPred) const {
    double mse = 0.0;
    if (yTrue.size() != yPred.size()) {
        throw std::invalid_argument("Vectors must be of the same size for mean squared error calculation.");
    }

    for (size_t i = 0; i < yTrue.size(); i++)
    {
        double error = yTrue[i] - yPred[i];
        mse += error * error;
    }

    mse /= yTrue.size();
    return mse;
}

/**
 * Predict the output for a given input using the learned parameters.
 * @param x: input feature
 * @return: predicted output
 */
double LinearRegression::predict(double x) const {
    return w * x + b;
}

/**
 * Fit the linear regression model to the training data.
 * @param x: input features
 * @param y: target values
 * @param epochs: number of iterations for training
 */
void LinearRegression::fit(const std::vector<double>& x, const std::vector<double>& y, int epochs, std::vector<double>& lossHistory) {
    if (x.size() != y.size()) {
        throw std::invalid_argument("Input and output vectors must be of the same size.");
    }

    // Initialize weights and bias
    w = 0.0;
    b = 0.0;
    const double n = static_cast<double>(x.size());

    // Gradient descent
    for (int epoch = 0; epoch < epochs; ++epoch) {
        double dwSum = 0.0;
        double dbSum = 0.0;
        double currentLoss = 0.0;

        for (size_t i = 0; i < x.size(); i++) {
            double yPred = predict(x[i]);
            double error = yPred - y[i];

            dwSum += (2 / n) * error * x[i];
            dbSum += (2 / n) * error;

            currentLoss += error * error; // accumulate loss for this epoch
        }

        w -= a * dwSum;
        b -= a * dbSum;
        // register loss history
        lossHistory.push_back(currentLoss / n);
    }
}

int main() {
    std::vector<double> X, y;
    std::default_random_engine generator(42);
    std::uniform_real_distribution<double> distribution(0.0, 2.0);
    std::normal_distribution<double> noise(0.0, 1.0);

    for (int i = 0; i < 100; ++i) {
        double x_val = distribution(generator);
        X.push_back(x_val);
        y.push_back(4.0 + 3.0 * x_val + noise(generator));
    }

    LinearRegression model;
    std::vector<double> loss_history;
    model.fit(X, y, 100, loss_history);

    std::ofstream results_file("results.csv");
    results_file << "X,y_true,y_pred\n";
    for(size_t i = 0; i < X.size(); ++i) {
        results_file << X[i] << "," << y[i] << "," << model.predict(X[i]) << "\n";
    }
    results_file.close();

    std::ofstream loss_file("loss_history.csv");
    loss_file << "loss\n";
    for(const auto& l : loss_history) {
        loss_file << l << "\n";
    }
    loss_file.close();
    
    std::cout << "C++ Linear Regression training finished. Results saved to CSV files." << std::endl;

    return 0;
}

File: linear_regression/linear_regression.py
import numpy as np
import matplotlib.pyplot as plt


"""
formula: y = Wx + b
where:
    y: target variable
    W: weights (coefficients)
    x: input features
    b: bias (intercept)
"""

class LinearRegression:
    def __init__(self, learning_rate):
        self.learning_rate = learning_rate
        self.weights = np.random.rand(1, 1)  # Initialize weights randomly
        self.bias = np.random.rand(1)  # Initialize bias randomly
        self.loss_history = []

    def hypothesis(self, x):   # Predict function model => y = Wx + b
        return np.dot(x, self.weights) + self.bias

    def mean_squared_error(self, y_true, y_pred):
        return np.mean((y_true - y_pred) ** 2)  # Squeared error loss (**: square operator)
    
    def fit(self, x, y, epochs):
        n_samples = x.shape[0]
        y = y.reshape(-1, 1)  # Ensure y is a column vector (because Wx + b expects column vector)
        
        for epoch in range(epochs):
            y_pred = self.hypothesis(x)
            loss = self.mean_squared_error(y, y_pred)
            self.loss_history.append(loss)

            error = y_pred - y  # Calculate error

            # Calculate gradients
            dW = (2 / n_samples) * np.dot(x.T, error)
            db = (2 / n_samples) * np.sum(error)

            # Update weights and bias
            self.weights -= self.learning_rate * dW
            self.bias -= self.learning_rate * db
    
        return self.loss_history

# Visualization of results
np.random.seed(42)
X = 2 * np.random.rand(100, 1)  # Generate random data, making feature data
y = 4 + 3 * X + np.random.randn(100, 1) # Add noise to the target variable  3 is Weight, 4 is Bias

model = LinearRegression(learning_rate=0.01)
losses = model.fit(X, y, epochs=100)

plt.style.use('seaborn-v0_8-whitegrid')

# Plotting the loss curve and regression line
fig, axes = plt.subplots(1, 2, figsize=(16, 6))

# Main title for the entire figure
fig.suptitle('Linear Regression Model Analysis', fontsize=16)


ax1 = axes[0] # Left side canvas
ax1.plot(losses, color='dodgerblue', linewidth=2)
ax1.set_title('Model Loss over Epochs', fontsize=14)
ax1.set_xlabel('Epochs')
ax1.set_ylabel('Mean Squared Error')


ax2 = axes[1] # Right side canvas
ax2.scatter(X, y, alpha=0.6, edgecolors='k', label='Data Points') 
ax2.plot(X, model.hypothesis(X), color='crimson', linewidth=3, label='Regression Line')
ax2.set_title('Regression Fit on Data', fontsize=14)
ax2.set_xlabel('X (Feature)')
ax2.set_ylabel('y (Target)')
ax2.legend()

# Adjust layout to prevent overlap with main title
plt.tight_layout(rect=[0, 0, 1, 0.96])
plt.savefig("linear_regression_analysis.png", dpi=150) # Save with higher resolution
plt.show()
plt.close()
